### DISCUSSION TITLE
How to create simple PostgreSQL procedure with return Table?

### DISCUSSION THREAD

**Discussion by AAlex-11** [supabase/supabase#30496] [Questions] [answered]
I need something like this
```
create or replace function path2()
returns table(
i integer,
id varchar,
name varchar,
mime varchar,
path varchar) as $$
BEGIN
WITH RECURSIVE x AS (
  SELECT i, parent, id, name, type, '' as path
	FROM entry WHERE name = 'JS-VBNET-2'
  UNION ALL
  SELECT e.i, e.parent, e.id, e.name, e.type, (x.path || '/' || e.name) as path
	FROM entry e, x WHERE x.id=e.parent
)
SELECT x.i, x.id, x.name, types.mime, x.path FROM x join types on types.i = x.type;
END;
$$ language plpgsql;
```
but I don't understand how I can create this function in this window, there are no parameter Table, I try to create with other return type but get errors

![VbNetSupport-11152024_204150_1](https://github.com/user-attachments/assets/83bd8e1f-0d7a-4ac6-a9b3-a12a0fec8527)

potentially I can create this SQL procedure in SQL editor, but in this case this will private procedure, I need public function what I can call from RPC

**Comment by GaryAustin1** [comment]
You should create in the SQL editor.
It will be exactly the same as the UI as far as access.

**Reply by AAlex-11** [reply]
Thanks. Can I call my function remotely with supabase.rpc('myFunction') ? This is my main goal.

And, unfortunately, this is only half of my problem. Second half that I can not convert my workable query to Function. This is query I made and what working perfectly

![VbNetSupport-11152024_211138_1](https://github.com/user-attachments/assets/aa4b7888-c15e-4838-a78f-96865f19c3ae)

I don't understand how in my function I need to define return table. 

![VbNetSupport-11152024_231104_1](https://github.com/user-attachments/assets/3bfecf37-ff88-4565-ac1e-3b3f28460af5)

**Reply by GaryAustin1** [reply, marked as answer]
Yes your function will work with the REST API rpc calls.  It is just like a function the UI creates to the database.

Your actual function table returning question is just Postgres so there should be lots of guides on the net for table returning functions or even the AI's do a good job at them.  I don't do table returning functions in my code so can't offer you advice on that.

**Reply by AAlex-11** [reply]
Thanks, @GaryAustin1

**Reply by AAlex-11** [reply]
correct function

```
CREATE
OR REPLACE FUNCTION PATH(enter TEXT, request TEXT) RETURNS TABLE (
  i INTEGER,
  id TEXT,
  NAME TEXT,
  types_mime TEXT,
  PATH TEXT
) AS $$
BEGIN
  WITH RECURSIVE x AS (
    SELECT i, parent, id, name, type, '' AS path
    FROM entry 
    WHERE name = enter
    UNION ALL
    SELECT e.i, e.parent, e.id, e.name, e.type, (x.path || '/' || e.name) AS path
    FROM entry e, x 
    WHERE x.id = e.parent
  )
  SELECT x.i, x.id, x.name, types.mime, x.path 
  FROM x 
  JOIN types ON types.i = x.type
  WHERE x.path = request;
END;
$$ LANGUAGE plpgsql;
```
